app/src/main/java/com/google/jetpackcamera/ui/PermissionsUi.kt

```diff
@@ -0,0 +1,202 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.jetpackcamera.ui
+
+import androidx.compose.foundation.background
+import androidx.compose.foundation.clickable
+import androidx.compose.foundation.layout.Arrangement
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.IntrinsicSize
+import androidx.compose.foundation.layout.Spacer
+import androidx.compose.foundation.layout.fillMaxHeight
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.foundation.layout.height
+import androidx.compose.foundation.layout.padding
+import androidx.compose.foundation.layout.size
+import androidx.compose.material3.Button
+import androidx.compose.material3.ButtonDefaults
+import androidx.compose.material3.Icon
+import androidx.compose.material3.MaterialTheme
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.painter.Painter
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.text.style.TextAlign
+import androidx.compose.ui.unit.dp
+import com.google.accompanist.permissions.ExperimentalPermissionsApi
+import com.google.accompanist.permissions.PermissionState
+import com.google.jetpackcamera.R
+
+@OptIn(ExperimentalPermissionsApi::class)
+@Composable
+fun CameraPermission(
+    modifier: Modifier = Modifier,
+    cameraPermissionState: PermissionState
+) {
+    PermissionTemplate(
+        modifier = modifier,
+        permissionState = cameraPermissionState,
+        painter = painterResource(id = R.drawable.photo_camera),
+        iconAccessibilityText = stringResource(id = R.string.image_accessibility_text),
+        title = stringResource(id = R.string.camera_permission_screen_title),
+        bodyText = stringResource(id = R.string.camera_permission_required_rationale),
+        requestButtonText = stringResource(R.string.request_permission)
+    )
+}
+
+@OptIn(ExperimentalPermissionsApi::class)
+@Composable
+fun PermissionTemplate(
+    modifier: Modifier = Modifier,
+    permissionState: PermissionState,
+    onSkipPermission: (() -> Unit)? = null,
+    painter: Painter,
+    iconAccessibilityText: String,
+    title: String,
+    bodyText: String,
+    requestButtonText: String
+) {
+    Column(
+        modifier = modifier.background(MaterialTheme.colorScheme.primary),
+        verticalArrangement = Arrangement.Bottom
+    ) {
+        // permission image / top half
```

Comment:
I think the comments you have are perfect places to extract out into sub Composables.